keps/sig-node/1592-pod-level-resource-limits/README.md

```diff
@@ -0,0 +1,439 @@
+pods not matching the `Guaranteed` or `Burstable` QoS definitions are placed in the `BestEffort` QoS level.
+
+This proposal suggests modifying these definitions the following: 
+
+A pod is considered `Guaranteed` if either of the following two conditions apply: 
+1. All of the containers included in the pod define memory and CPU resource requests equal to resource limits
+1. The pod includes a pod-level memory and CPU resource request which is equal to the resource limit, even if one or more of the included containers doesn't define either a CPU or a memory request or limit.
+
+A pod is considered `Burstable` if at least one container has a resource request, or the pod itself has a resource request. As before this proposal, setting
+a limit without a request causes Kubernetes to use the limit as a request.
+
+A pod is considered `BestEffort` if there are not resource request or limits settings on any container or on the pod level itself.
+
+If a developer included a request and limit definition at the pod level, the developer's intent was for Kubernetes to attempt to guarantee resources 
+as a whole, for all the containers included in the pod. This should apply even if the developer is unable to preemptively split the resources correctly
+between all the containers contained in the pod. 
+
+Not all resource limits are treated in the same way by the underlying Linux cgroup controllers.  For example, if a memory pod level limit request is smaller 
+than the sum of limit for all the containers, then the limit requested by the containers is unreachable. It becomes impossible for all of the containers to 
+reach their requested limit, since the sum of available memory becomes the pod-level limit. Note that even though setting `pod.limit[memory] < sum(pod.containers[*].limit[memory])`
+would not allow all of the containers to allocate memory up to their limit, this is not considered an error by this proposal (though a warning should be emitted).
+If it were considered an error, then any admission controller that injects containers to pods with defined container-level limits might cause the pod to fail in unexpected ways. 
+
+Conversely if `pod.limit[memory] > sum(pod.containers[*].limit[memory])` then the excess is not relevant since the individual containers 
+can never use more than their defined limit, and even if all of the containers use the maximum allowed by their defined memory limit, it can never amount
+to the pod-level limit. Setting such a limit on the pod level does not _reserve_ any memory to the pod and will not negatively affect the rest of the 
+workloads running on the node. Therefore this is not considered an error. 
+
+Compare this to the CPU resource, where the underlying Linux cgroup treats the limit as a ratio relative to the cgroup one level up in the hierarchy. In this
+case setting the pod-level CPU limit will limit that pod to shares equal to the requested amount of CPU shares. Once the pod limit is set, then the limit on
```

Comment:
This says "shares" here, but I thought shares were how `requests` are implemented, not `limits`?

Reply:
@thockin Agree, I would make that change as well.

Relevant lithography from the [Resource Management doc](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#how-pods-with-resource-limits-are-run)

> The CPU limit defines a hard ceiling on how much CPU time that the container can use. During each scheduling interval (time slice), the Linux kernel checks to see if this limit is exceeded; if so, the kernel waits before allowing that cgroup to resume execution.
> The CPU request typically defines a weighting. If several different containers (cgroups) want to run on a contended system, workloads with larger CPU requests are allocated more CPU time than workloads with small requests.